pkg/querier/batch/batch.go

```diff
@@ -55,17 +56,37 @@ type iterator interface {
 	Err() error
 }
 
+var genericChunkSlicePool zeropool.Pool[[]GenericChunk]
+
+func getGenericChunkSlice(n int) []GenericChunk {
+	sn := genericChunkSlicePool.Get()
+	if sn == nil || cap(sn) < n {
```

Comment:
Would some kind of pool with different buckets for different ranges of `n` be better here, to avoid the pool only ever having large slices?

Reply:
+1 in a multi tenant installation I assume you get all kinds of different size queries so you'll end up increasing the memory footprint over time by having more and more large slices in the pool. Maybe measure first with a histogram in some busy environment to get a good idea on buckets?

Note: I've checked that this is invoked after applying the max chunks per query limit.

Reply:
you can maybe use `SlabPool` ([code](https://github.com/grafana/mimir/blob/942e16cfc8699ed2ea35aad439d05c730560a556/pkg/util/pool/pool.go#L25-L36)) which can pack together smaller slices into one big slice. We used to have bucketed pool, but it was unused and removed in https://github.com/grafana/mimir/pull/4996

Reply:
we use this code in the store-gateways. The way we size slabs is to account for the worst-case or p99 case of number of chunks per series. This is [an example](https://github.com/grafana/mimir/blob/14080d085561056f7ef8aa70a5641cb403491c3b/pkg/storegateway/series_chunks.go#L26-L31) for slices of chunks.

So if a series can have at most 300 chunks, have a slab size of 300. If all series have fewer chunks, e.g. 3, then one slab can fit 100 series. The overhead isn't huge, but you also get reduced number of overall allocations and maybe even better CPU cache locality since sequential slices are sequential in memory.